src/Storages/S3Queue/S3QueueFilesMetadata.cpp

```diff
@@ -28,326 +42,875 @@ namespace
+                                                         zkutil::CreateMode::Persistent));
+        }
+        else
+        {
+            requests.push_back(zkutil::makeSetRequest(zookeeper_failed_path / node_name_with_retriable_suffix,
+                                                      node_metadata.toString(),
+                                                      stat.version));
+        }
+        Coordination::Responses responses;
+        auto code = zk_client->tryMulti(requests, responses);
+        if (code == Coordination::Error::ZOK)
+            return;
+
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Failed to set file as failed");
+    }
 }
 
-void S3QueueFilesMetadata::removeProcessingFile(const String & file_path)
+S3QueueFilesMetadata::ProcessingNodeHolder::ProcessingNodeHolder(
+    const std::string & processing_id_,
+    const std::string & path_,
+    const std::string & zk_node_path_,
+    zkutil::ZooKeeperPtr zk_client_)
+    : zk_client(zk_client_)
+    , path(path_)
+    , zk_node_path(zk_node_path_)
+    , processing_id(processing_id_)
+    , log(&Poco::Logger::get("ProcessingNodeHolder"))
 {
-    auto zookeeper = storage->getZooKeeper();
-    String processing_files;
-    zookeeper->tryGet(zookeeper_processing_path, processing_files);
-
-    S3QueueProcessingCollection processing_collection;
-    processing_collection.parse(processing_files);
-    processing_collection.remove(file_path);
-    zookeeper->set(zookeeper_processing_path, processing_collection.toString());
 }
 
-S3QueueFilesMetadata::S3FilesCollection S3QueueFilesMetadata::getUnorderedProcessedFiles()
+S3QueueFilesMetadata::ProcessingNodeHolder::~ProcessingNodeHolder()
 {
-    auto zookeeper = storage->getZooKeeper();
-    S3QueueProcessedCollection processed_collection(max_set_size, max_set_age_sec);
-    processed_collection.parse(zookeeper->get(zookeeper_processed_path));
-    return processed_collection.getFileNames();
+    if (!removed)
+        remove();
 }
 
-S3QueueFilesMetadata::S3FilesCollection S3QueueFilesMetadata::getProcessedFailedAndProcessingFiles()
+bool S3QueueFilesMetadata::ProcessingNodeHolder::remove(Coordination::Requests * requests, Coordination::Responses * responses)
 {
-    S3FilesCollection processed_and_failed_files = getFailedFiles();
-    switch (mode)
+    if (removed)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Processing node is already removed");
+
+    LOG_TEST(log, "Removing processing node {} ({})", zk_node_path, path);
+
+    try
     {
-        case S3QueueMode::UNORDERED:
+        if (!zk_client->expired())
         {
-            processed_and_failed_files.merge(getUnorderedProcessedFiles());
-            break;
+            /// Is is possible that we created an ephemeral processing node
+            /// but session expired and someone other created an ephemeral processing node.
+            /// To avoid deleting this new node, check processing_id.
```

Comment:
I didn't quite understand when it's possible. Can you explain with more details?

Reply:
Let' say we have `clickhouse-server1` and `clickhouse-server2`, create `S3Queue` table on both servers pointing to the same keeper path. Let's say server 1 started processing `file1`, but got keeper session expired before finishing. Expired session leads to expiration of ephemeral "processing" node. Server 2 sees that there is no "processing" nor "failed" nor "processed" node for `file1`, so it will create a "processing" node itself. Then server1 restores keeper connection, it knows from memory that it was processing `file1` and did not finish. So it try to finalize it with either "failed" or "finished" state. But it will be incorrect if the described scenario.

Reply:
So we need to think about deduplication here. If I understood correctly, the solution is to modify read offset in zk (or read number of rows, see my comments below) after each read block in source, so if we cannot modify it because of session expired, we don't insert this block. Right?

Reply:
Yeah, also there is a case that if we will update the counter in keeper after we read the block and before we pushed the block to mv, then there could be an exception during push and we get incorrect state in keeper, then to make things even worse - keeper session expires - so we cannot fix keeper state without a potential race with another server having s3queue starting processing the same file, so it is again not straightforward what to do here.

So (at least for now) I'd better document that the user is strongly recommended to use destination table of S3Queue MV with table engine which supports deduplication. Then occasional duplicated is not a problem.